src/preserve.rs

```diff
@@ -0,0 +1,75 @@
+use serde::{
+    de::{self, Visitor},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::{fmt, sync::Mutex};
+use wasm_bindgen::JsValue;
+
+/// This type is used to preserve a [`JsValue`] when serializing and deserializing.
+///
+/// ```rust
+/// #[derive(Serialize, Deserialize)]
+/// struct MyStruct {
+///    // works with objects from wasm-bindgen; they just need to implement From<JsValue> and Into<JsValue>
+///    my_value: PreserveJsValue<ObjectFromWasmBindgen>,
+///    // and with raw JsValue values
+///    my_other_value: PreserveJsValue<JsValue>,
+/// }
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub struct PreserveJsValue<T: From<JsValue> + Into<JsValue> + Clone>(pub T);
+impl<T: From<JsValue> + Into<JsValue> + Clone> From<T> for PreserveJsValue<T> {
+    fn from(value: T) -> Self {
+        Self(value)
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct JsValueKeeper(pub JsValue);
+unsafe impl Send for JsValueKeeper {}
+unsafe impl Sync for JsValueKeeper {}
+
+pub(crate) static NEXT_PRESERVE: Mutex<Option<JsValueKeeper>> = Mutex::new(None);
+
+impl<T: From<JsValue> + Into<JsValue> + Clone> Serialize for PreserveJsValue<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        NEXT_PRESERVE
+            .lock()
+            .unwrap()
+            .replace(JsValueKeeper(self.0.clone().into()));
+        serializer.serialize_i64(0)
+    }
+}
+
+impl<'de, T: From<JsValue> + Into<JsValue> + Clone> Deserialize<'de> for PreserveJsValue<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct JsValueVisitor<T: From<JsValue> + Into<JsValue> + Clone>(
+            std::marker::PhantomData<T>,
+        );
```

Comment:
Is there any benefit to this type being generic? Why not just use the outer `T` that's already bound?

Reply:
Hm iirc you can't bind to outer generics like that in Rust? Could be wrong though